dotnet/src/SemanticKernel/AI/OpenAI/Clients/RetryHandler.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.SemanticKernel.AI.OpenAI.Clients;
+
+internal class RetryHandler: DelegatingHandler
+{
+    private const string Retry_After = "Retry-After";
+    private const int MaxRetries = 10;
+
+    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+    {
+        int retryCount = 0;
+
+        while (true)
+        {
+            // Throw an exception if we've requested to cancel the operation
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Issue the request
+            HttpResponseMessage response = await base.SendAsync(request, cancellationToken);
+
+            // If the request does not require a retry then we're done
+            if (!ShouldRetry(response.StatusCode))
+            {
+                return response;
+            }
+
+            // Drain response content to free connections. Need to perform this
+            // before retry attempt and before the TooManyRetries ServiceException.
+            if (response.Content != null)
+            {                
+#if NET5_0_OR_GREATER
+                await response.Content.ReadAsByteArrayAsync(cancellationToken).ConfigureAwait(false);
+#else
+                await response.Content.ReadAsByteArrayAsync().ConfigureAwait(false);
+#endif
+            }
+
+            // Safety measure to not keep retrying forever
+            if (retryCount >= MaxRetries)
+            {
+                throw new AIException(AIException.ErrorCodes.UnknownError,
+                    $"Request reached it's max retry count of {retryCount}");
```

Comment:
We can return response here instead of throwing the exception. The response has more information (status code, headers, etc) than the exception which might be useful for the other potential handlers in the handlers chain if they need to act upon it. And it seems the [OpenAI Service Client class](https://github.com/microsoft/semantic-kernel/blob/main/dotnet/src/SemanticKernel/AI/OpenAI/Clients/OpenAIClientAbstract.cs#:~:text=switch%20(response.StatusCode)) can successfully handle it by checking if it’s successful or not and mapping one of the HTTP Status codes the handler handles to the AIException.